node/auth/authenticator.go

```diff
@@ -0,0 +1,172 @@
+
+type requestAuthenticator struct {
+	// chainReader is used to read the chain state.
+	chainReader core.Reader
+
+	// keyCache is used to cache the public keys of dispersers.
+	keyCache *lru.Cache[uint32, *keyWithTimeout]
+
+	// keyTimeoutDuration is the duration for which a key is cached. After this duration, the key should be
+	// reloaded from the chain state in case the key has been changed.
+	keyTimeoutDuration time.Duration
+
+	// authenticatedDispersers is a set of disperser addresses that have been recently authenticated, mapped
+	// to the time when that cached authentication will expire.
+	authenticatedDispersers *lru.Cache[string, time.Time]
+
+	// authenticationTimeoutDuration is the duration for which an auth is valid.
+	// If this is zero, then auth saving is disabled, and each request will be authenticated independently.
+	authenticationTimeoutDuration time.Duration
+}
+
+// NewRequestAuthenticator creates a new RequestAuthenticator.
+func NewRequestAuthenticator(
+	ctx context.Context,
+	chainReader core.Reader,
+	keyCacheSize int,
+	keyTimeoutDuration time.Duration,
+	authenticationTimeoutDuration time.Duration,
+	now time.Time) (RequestAuthenticator, error) {
+
+	keyCache, err := lru.New[uint32, *keyWithTimeout](keyCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create key cache: %w", err)
+	}
+
+	authenticatedDispersers, err := lru.New[string, time.Time](keyCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create authenticated dispersers cache: %w", err)
+	}
+
+	authenticator := &requestAuthenticator{
+		chainReader:                   chainReader,
+		keyCache:                      keyCache,
+		keyTimeoutDuration:            keyTimeoutDuration,
+		authenticatedDispersers:       authenticatedDispersers,
+		authenticationTimeoutDuration: authenticationTimeoutDuration,
+	}
+
+	err = authenticator.preloadCache(ctx, now)
+	if err != nil {
+		return nil, fmt.Errorf("failed to preload cache: %w", err)
+	}
+
+	return authenticator, nil
+}
+
+func (a *requestAuthenticator) preloadCache(ctx context.Context, now time.Time) error {
+	// this will need to be updated for decentralized dispersers
+	_, err := a.getDisperserKey(ctx, now, EigenLabsDisperserID)
+	if err != nil {
+		return fmt.Errorf("failed to get operator key: %w", err)
+	}
+
+	return nil
+}
+
+func (a *requestAuthenticator) AuthenticateStoreChunksRequest(
+	ctx context.Context,
+	origin string,
+	request *grpc.StoreChunksRequest,
+	now time.Time) error {
+
+	if a.isAuthenticationStillValid(now, origin) {
+		// We've recently authenticated this client. Do not authenticate again for a while.
+		return nil
+	}
+
+	key, err := a.getDisperserKey(ctx, now, request.DisperserID)
+	if err != nil {
+		return fmt.Errorf("failed to get operator key: %w", err)
+	}
+
+	err = VerifyStoreChunksRequest(*key, request)
+	if err != nil {
+		return fmt.Errorf("failed to verify request: %w", err)
+	}
+
+	a.saveAuthenticationResult(now, origin)
```

Comment:
nit: `cacheAuthenticationResult`, "save" feels like it's persisting the data which it's not

Reply:
renamed